### PR TITLE
Standardized-mod-format specification

### PR DESCRIPTION
**Pull request by 20kdc**
DRAFT!
DO NOT MERGE YET!

This particular PR is NOT fully-worked-out yet.

Unlike the other PRs, it absolutely *requires* discussion from modloader authors to be made usable.
This means @2767mr and @ac2pic.

### CODE REVIEW

File: proposals/1/standardized-mod-format.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
+# Standardized Mod Format (DRAFT!)
+
+## Description Of Issue
+
+Lack of a clear specification between modloaders on a common supported format.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: Please specify which format is described (package.json?)

Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: Maybe add some more reasoning

File: proposals/1/standardized-mod-format.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+## Solution
+
+A specification which exactly defines specific terms, and does not define legacy formats from the previous era.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md:11]: Why should legacy formats not be described?

Comment (2767mr) [inline proposals/1/standardized-mod-format.md:11]: This does not specify what this spec is actually about (package.json)

File: proposals/1/standardized-mod-format.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+### Advantages
+
+- Modloaders and modders know what is supported everywhere.
+
+- Less likelihood to use outdated mechanisms from the old era
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md:18]: Loading mechanisms from the old era wont be able to follow this spec because it hasn't existed then

File: proposals/1/standardized-mod-format.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+- APIs defined in other specifications may gain reference implementations using this specification - these can then become the only implementation of those APIs, preventing fragmentation.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: This is the case for all specs. A single forced implementation of a spec is usually a bad idea.

File: proposals/1/standardized-mod-format.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+### Disadvantages
+
+- Per-modloader extensions not covered, theoretically limiting features for more portable mods.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: Why excatly are specifically portable mods mentioned?

Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: The spec needs to include the handling of features unique to certain modloaders.

File: proposals/1/standardized-mod-format.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+### Regarding Implementation
+
+It is important to note that some features of this specification may, themselves, be implemented within mods.
+
+This is fine - the implementation of this specification is then a matter of "that modloader + these mods".
+
+This in particular applies to how CCLoader handles patch files via `simplify`.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md:30]: Maybe reduce this section to "It doesn't matter if this spec is implemented using only the modloader or with mods". You could also remove this section since there are no API's or restrictions defined in this spec.

File: proposals/1/standardized-mod-format.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+## Mod Format
+
+A mod is in a directory. The directory's name is the ID of the mod.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: This would directly go against the upcoming packaged mod proposal

File: proposals/1/standardized-mod-format.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+The directory then contains a file called "package.json".
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: Are subdirectories included? CCLoader does not include them currently.

File: proposals/1/standardized-mod-format.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+### The package.json format
+
+The package.json format describes a mod.
+
+It has the following elements:
+
+```
+{
+ "version": <Semver-compliant version string>,
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md:49]: Should the version be required?

File: proposals/1/standardized-mod-format.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+ "ccmodDependencies": <Optional Object - keys are mod IDs, values are semver-compliant version constraints - This is used between different mods>,
+ "dependencies": <
+  Optional Object - due to compatibility, this is a bit awkward
+  If ccmodDependencies is provided, keys are Node modules, values are semver-compliant version constraints
+  If ccmodDependencies is NOT provided, this key is effectively renamed to it - meaning keys are mod IDs, not Node modules.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md:58]: Maybe rephrase this from "rename" to "handled as"

File: proposals/1/standardized-mod-format.md [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+ >,
+
+ "name": <Optional human-readable string - otherwise, the directory name is used>,
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: Should this be the human-readable string? The npm spec specifies strict limits on the package name.

File: proposals/1/standardized-mod-format.md [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+ "description": <Optional human-readable string>,
+
+ "preload": <Optional filename of a script>,
+ "postload": <Optional filename of a script>,
+ "prestart": <Optional filename of a script>,
+ "main": <Optional filename of a script. For example, "mod.js" loads the file "mod.js" within the mod folder>,
+
+ "usesRequire": <Optional boolean - if true, this mod uses the 'require' function. Defaults to true>,
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: What happens if this is set to false?

File: proposals/1/standardized-mod-format.md [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+ "assets": <
+  Optional array of strings, such as "media/icons.png".
+  An asset ending in ".json.patch" is a patch file - see Patch Files.
+  If not provided, assets may be automatically guessed or completely ignored.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: Keeping this up to date is a real pain. Should this be really ignored if missing? Maybe add a `true` or dirname option to automatically infer assets.

File: proposals/1/standardized-mod-format.md [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+  (NOTE! THIS PART CONFLICTS WITH CURRENT BEHAVIOR - CCLOADER RIGHT NOW WILL ALWAYS GUESS ASSETS, EVEN IF PROVIDED.)
+ >
+}
+```
+
+### Patch Files
+
+Patch files have the `.json.patch` extension, and patch the relevant `.json` asset.
+
+Patch files are a tree of objects.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: This should mention that patch files are stored as json

File: proposals/1/standardized-mod-format.md [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+### The package.json format
+
+The package.json format describes a mod.
+
+It has the following elements:
+
+```
+{
+ "version": <Semver-compliant version string>,
+ "ccmodDependencies": <Optional Object - keys are mod IDs, values are semver-compliant version constraints - This is used between different mods>,
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: Since mod IDs are the folder names these would break if you rename the mod folder. This can happen due to installation errors

File: proposals/1/standardized-mod-format.md [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+
+The routine is:
+
+For each key in the current patch object...
+
+1. If the current target's value for the key is undefined, set it to the value for the key in the patch object.
+2. Else, if the patch object's value for the key is an Object (and not a subclass - think boxed type coercion), apply the algorithm for (the current target's value for the key, the patch object's value for the key).
+3. Else, set the current target's value for the key to the value for the key in the patch object.
+
+Note from the future: If the `patch-steps` proposal becomes part of CLS, it then also becomes a valid patch format.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: Future proposals should not be mentioned in any spec

File: proposals/1/standardized-mod-format.md [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+
+### Virtual Dependencies
+
+There are two "virtual dependencies" known. These do not have mod IDs.
+
+"ccloader": Forces a dependency on CCLoader. A mod that uses this is not compliant.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: Maybe generalise this to "name of modloader"

File: proposals/1/standardized-mod-format.md [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+
+Within a given 'phase', mods must be loaded in dependency order.
+
+The behavior for a circular dependency situation is undefined.
+
+The 'phases' are:
+
+1. `preload`. This is executed before game.compiled.js is executed.
+2. `postload`. This is executed after game.compiled.js is executed, but before window.startCrossCode.
+3. `prestart` (Name subject to change) This occurs directly before `ig.main`, which ensures the whole game has loaded and effectively gives the code free reign over Cubic Impact structures.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: Please further clarify "directly before `ig.main`"

File: proposals/1/standardized-mod-format.md [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+4. `main`. This is executed at any time after the game has completely loaded and is running, i.e. `ig.ready` is true, `ig.game` exists, etc.
+
+---
+
+No implementations are registered here at this time because this spec is still way too far into flux for anything to be settled as "the" implementation.
```
Comment (2767mr) [inline proposals/1/standardized-mod-format.md]: This should be a bit rephrased when it is accepted as spec

**Review by 2767mr — COMMENTED** [review]
Some general comments. Most of them are open for discussion.

File: proposals/1/standardized-mod-format.md (hunk 2 above)
Comment (20kdc) [reply proposals/1/standardized-mod-format.md:11]: Because it'd either implicitly require support for them or be kind of a waste of space...

File: proposals/1/standardized-mod-format.md (hunk 5 above)
Comment (20kdc) [reply proposals/1/standardized-mod-format.md]: I'm thinking "per-modloader extension optionally used by a portable mod" in regards to this disadvantage.

File: proposals/1/standardized-mod-format.md (hunk 13 above)
Comment (20kdc) [reply proposals/1/standardized-mod-format.md]: Impossible if non-local.
If the specification should or should not support non-local is up for debate, though.

File: proposals/1/standardized-mod-format.md (hunk 15 above)
Comment (20kdc) [reply proposals/1/standardized-mod-format.md]: Apparently I misread how CCLoader was handling this - I assumed that it was using folder names and wrote this this way for compatibility. Will fix.

Comment (2767mr) [reply proposals/1/standardized-mod-format.md]: CCLoader will use folder names if no name is provided in the package.json

File: proposals/1/standardized-mod-format.md (hunk 7 above)
Comment (20kdc) [reply proposals/1/standardized-mod-format.md]: Ignoring the (now revised) "directory name == ID of mod" error...
I'm assuming that the directory in that case is virtual, but does conceptually "exist".

File: proposals/1/standardized-mod-format.md (hunk 9 above)
Comment (20kdc) [reply proposals/1/standardized-mod-format.md:49]: It'd make the dependency situation more complicated if a mod neglected to provide a version key.
This is only the specification for a conformant standardized mod, so modloaders are free to accept-but-warn or such.

**Comment by 20kdc** [comment]
Things still to discuss:
1. Finalization of "prestart" phase (& name) (EDIT: And *possibly* if it is to exist, but I think I've already noted in the proposal as to why that should either be a 'yes' or be effectively emulated via other methods)
2. Is "preload" & "postload" necessary to be defined *other* than as reference points?
3. Anything I've missed

EDIT:
4. With assets now forcefully implied, this means standardized-mods won't work in a browser context.
5. I'm not sure what to do about module.

File: proposals/1/standardized-mod-format.md (hunk 2 above)
Comment (2767mr) [reply proposals/1/standardized-mod-format.md:11]: Defining legacy formats as optional would allow modloaders to support old mods

File: proposals/1/standardized-mod-format.md [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+  This is used between different mods to establish load order & required other mods.
+  Note that some entries in here may not actually be mods - see Virtual Dependencies for further information.
+ >,
+ "dependencies": <
+  Optional Object - due to compatibility, this is a bit awkward
+  If ccmodDependencies is provided, keys are Node modules, values are semver-compliant version constraints
+  If ccmodDependencies is NOT provided, this key is effectively renamed to it - meaning keys are mod IDs, not Node modules.
+ >,
+
+ "description": <Optional human-readable string>,
```
Comment (streetclaw) [inline proposals/1/standardized-mod-format.md:61]: Request for adding fields `author` (optional; maintainer/main author) and `contributors` (optional, array). Reference: https://docs.npmjs.com/files/package.json#people-fields-author-contributors

File: proposals/1/standardized-mod-format.md (hunk 2 above)
Comment (20kdc) [reply proposals/1/standardized-mod-format.md:11]: This specification does not state that modloaders can't have their own support mechanisms if they so choose.

File: proposals/1/standardized-mod-format.md (hunk 20 above)
Comment (20kdc) [reply proposals/1/standardized-mod-format.md:61]: Resolved with b5d0e83 clarifying that these are allowed without explicitly defining their format.

File: proposals/1/standardized-mod-format.md (hunk 2 above)
Comment (20kdc) [reply proposals/1/standardized-mod-format.md:11]: Resolved with b5d0e83 clarifying that these are allowed without explicitly defining their format.

**Comment by 20kdc** [comment]
The spec's usable loading phases have been reduced to `postload` and `main` since 2767mr doesn't seem to like the idea of implementing `prestart` (it's emulatable anyway so *shrugs*)

EDIT: As such, the TODO list:

1. With assets now forcefully implied, this means standardized-mods won't work in a browser context.
(If this is good/bad... IDK!)
2. I'm not sure what to do about module.
3. Anything I've missed

File: proposals/1/standardized-mod-format.md (hunk 20 above)
Comment (streetclaw) [reply proposals/1/standardized-mod-format.md:61]: I would prefer to have `author` and `contributors` standardized. So every tool/loader can use this data if needed.

**Comment by ghost** [comment]
Before @2767mr comments, I requested that main not be a phase because it supports bad practices.

**Comment by 20kdc** [comment]
The only thing really left to discuss here is probably if the conformance tests should be included.
Given all the changes, you *should* re-read the re-re-re-re-re-re-revised spec before stamping approval.

**Review by ghost — APPROVED** [review]

**Review by 2767mr — APPROVED** [review]
This should be enough for the proposal to be merged.
